src/gcp.rs

```diff
@@ -194,14 +197,22 @@ impl GcpAuthenticator {
     async fn token_regenerator(self, sender: watch::Sender<()>) {
         match self {
             Self::Credentials(inner) => {
-                let period =
-                    Duration::from_secs(inner.token.read().unwrap().expires_in() as u64 / 2);
-                let mut interval = tokio::time::interval_at(Instant::now() + period, period);
+                let expires_in = inner.token.read().unwrap().expires_in() as u64;
+                let next_refresh = sleep(Duration::from_secs(expires_in - METADATA_TOKEN_EXPIRY_MARGIN_SECS));
+                tokio::pin!(next_refresh);
                 loop {
-                    interval.tick().await;
                     debug!("Renewing GCP authentication token.");
                     match inner.regenerate_token().await {
-                        Ok(()) => sender.send_replace(()),
+                        Ok(()) => {
+                            sender.send_replace(());
+                            tokio::select! {
+                                () = &mut next_refresh => {
+                                    let expires_in = inner.token.read().unwrap().expires_in() as u64;
+                                    let next_deadline = Instant::now() + Duration::from_secs(expires_in - METADATA_TOKEN_EXPIRY_MARGIN_SECS);
+                                    next_refresh.as_mut().reset(next_deadline);
+                                },
+                            }
```

Comment:
One potential issue here is that we're only waiting in the `Ok` branch after regenerating a token. If we get an error, it looks like this will busy loop and potentially cause problems if there is some kind of upstream outage, etc.

Ideally we would keep the original `await` at the very beginning of the loop, though I can see there is some complexity in that we can only reset the sleep when we get a new token with a new expiry. Maybe we can have a separate amount of time that we reset to in the case of an error? This would help us handle errors more gracefully.

Separately, having a `tokio::select!` with only one branch should not be necessary, as it should be equivalent to `next_refresh.await`.

Reply:
Yes, I see I missed adding the original `await` and that brought me into a rabbit hole. Thanks for your comments!

I've updated the code and, as suggested, added a separate deadline for when an Error is encountered.